File: lab11/lab11-2.py
class Solution:
    def checkValidString(self, s):
        stack = [(0, 0)]  
        while stack:
            index, balance = stack.pop()
            if balance < 0:
                continue
            if index == len(s) and balance == 0:
                return True
            if index < len(s):
                if s[index] == '(':
                    stack.append((index + 1, balance + 1))  
                elif s[index] == ')':
                    stack.append((index + 1, balance - 1))  
                else:  # '*'
                    stack.append((index + 1, balance + 1))  
                    stack.append((index + 1, balance - 1))  
                    stack.append((index + 1, balance))      
        return False
solution = Solution()
print(solution.checkValidString("()"))    
print(solution.checkValidString("(*)"))   
print(solution.checkValidString("(*))")) 

File: lab11/lab11-2.c
#include <stdio.h>
#include <stdbool.h>

bool checkValidString(char *s) {
    int stack_size = 1000; 
    int stack[stack_size][2]; 
    int top = -1;
    stack[++top][0] = 0;
    stack[top][1] = 0;

    while (top >= 0) {
        int index = stack[top][0];
        int balance = stack[top][1];
        top--;

        if (balance < 0) {
            continue;
        }
        if (index == strlen(s) && balance == 0) {
            return true;
        }
        if (index < strlen(s)) {
            if (s[index] == '(') {
                stack[++top][0] = index + 1;
                stack[top][1] = balance + 1;
            } else if (s[index] == ')') {
                stack[++top][0] = index + 1;
                stack[top][1] = balance - 1;
            } else {  
                stack[++top][0] = index + 1;
                stack[top][1] = balance + 1;

                stack[++top][0] = index + 1;
                stack[top][1] = balance - 1;

                stack[++top][0] = index + 1;
                stack[top][1] = balance;
            }
        }
    }
    return false;
}

int main() {
    printf("%d\n", checkValidString("()"));       // Should print 1 (true)
    printf("%d\n", checkValidString("(*)"));      // Should print 1 (true)
    printf("%d\n", checkValidString("(*))"));     // Should print 1 (true)

    return 0;
}